guardian/managers.py

```diff
@@ -26,7 +26,23 @@ def is_generic(self):
         except models.fields.FieldDoesNotExist:
             return False
 
-    def assign_perm(self, perm, user_or_group, obj):
+    def _perm_kwargs(self, permission, user_or_group, obj, ctype=None, user_group_object=None):
+        if ctype is None:
+            ctype = get_content_type(obj)
+        kwargs = {
+            'permission': permission,
+            self.user_or_group_field: user_or_group,
+        }
+        if self.is_generic():
+            kwargs['content_type'] = ctype
+            kwargs['object_pk'] = obj.pk
+        else:
+            kwargs['content_object'] = obj
+        if user_group_object:
+            kwargs['user_group_object'] = user_group_object
+        return kwargs
+
+    def assign_perm(self, perm, user_or_group, obj, user_group_object=None):
         """
         Assigns permission with given ``perm`` for an instance ``obj`` and
```

Comment:
I'll rename the class UserGroupObject to Origin and variables user_group_object to origin